### ISSUE TITLE
Back end exception during validation if required field is unmapped

### ISSUE THREAD

**Issue by maxxxxxdlp** [specify/specify7#777]
`Accession Agent` didn't have required relationship-field `Agent` mapped. When modifying the table, autovalidation produced the following exception:

```
IntegrityError at /api/workbench/validate_row/1/
(1048, "Column 'AgentID' cannot be null")

Request Method: POST
Request URL: http://localhost/api/workbench/validate_row/1/
Django Version: 2.2.10
Python Executable: /opt/specify7/ve/bin/python
Python Version: 3.6.9
Python Path: ['/opt/specify7', '/opt/specify7', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/opt/specify7/ve/lib/python3.6/site-packages']
Server time: Fri, 5 Feb 2021 18:22:23 -0600
Installed Applications:
('django.contrib.sessions',
 'django.contrib.staticfiles',
 'django.contrib.contenttypes',
 'django.contrib.auth',
 'specifyweb.specify',
 'specifyweb.stored_queries',
 'specifyweb.businessrules',
 'specifyweb.express_search',
 'specifyweb.context',
 'specifyweb.attachment_gw',
 'specifyweb.frontend',
 'specifyweb.barvis',
 'specifyweb.report_runner',
 'specifyweb.interactions',
 'specifyweb.workbench',
 'specifyweb.notifications',
 'specifyweb.export',
 'specifyweb.raven_placeholder',
 'django_jsonfield_backport')
Installed Middleware:
['django.middleware.gzip.GZipMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'specifyweb.context.middleware.ContextMiddleware']


Traceback:

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/backends/mysql/base.py" in execute
  71.             return self.cursor.execute(query, args)

File "/opt/specify7/ve/lib/python3.6/site-packages/MySQLdb/cursors.py" in execute
  250.             self.errorhandler(self, exc, value)

File "/opt/specify7/ve/lib/python3.6/site-packages/MySQLdb/connections.py" in defaulterrorhandler
  50.         raise errorvalue

File "/opt/specify7/ve/lib/python3.6/site-packages/MySQLdb/cursors.py" in execute
  247.             res = self._query(query)

File "/opt/specify7/ve/lib/python3.6/site-packages/MySQLdb/cursors.py" in _query
  412.         rowcount = self._do_query(q)

File "/opt/specify7/ve/lib/python3.6/site-packages/MySQLdb/cursors.py" in _do_query
  375.         db.query(q)

File "/opt/specify7/ve/lib/python3.6/site-packages/MySQLdb/connections.py" in query
  276.             _mysql.connection.query(self, query)

During handling of the above exception ((1048, "Column 'AgentID' cannot be null")), another exception occurred:

File "/opt/specify7/ve/lib/python3.6/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/opt/specify7/specifyweb/specify/views.py" in wrapped
  20.         return view(request, *args, **kwargs)

File "/opt/specify7/specifyweb/specify/views.py" in wrapped
  32.         return view(request, *args, **kwargs)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/views/decorators/http.py" in inner
  40.             return func(request, *args, **kwargs)

File "/opt/specify7/specifyweb/workbench/views.py" in validate_row
  288.         result = uploader.validate_row(collection, upload_plan, request.POST)

File "/opt/specify7/specifyweb/workbench/upload/upload.py" in validate_row
  137.                 result = UploadResult(bind_result, {}, {}) if isinstance(bind_result, ParseFailures) else bind_result.process_row()

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in process_row
  155.         return self._handle_row(force_upload=False)

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in _handle_row
  196.         return self._do_upload(model, toOneResults, attrs, info)

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in _do_upload
  257.             for fieldname, records in self.toMany.items()

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in <dictcomp>
  257.             for fieldname, records in self.toMany.items()

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in _upload_to_manys
  317.         for record in records

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in <listcomp>
  317.         for record in records

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in force_upload_row
  158.         return self._handle_row(force_upload=True)

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in _handle_row
  196.         return self._do_upload(model, toOneResults, attrs, info)

File "/opt/specify7/specifyweb/workbench/upload/upload_table.py" in _do_upload
  250.             uploaded = model.objects.create(**attrs, **self.scopingAttrs, **self.static)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/manager.py" in manager_method
  82.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/query.py" in create
  422.         obj.save(force_insert=True, using=self.db)

File "/opt/specify7/specifyweb/specify/build_models.py" in save
  55.             return super(model, self).save(*args, **kwargs)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/base.py" in save
  741.                        force_update=force_update, update_fields=update_fields)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/base.py" in save_base
  779.                 force_update, using, update_fields,

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/base.py" in _save_table
  870.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/base.py" in _do_insert
  908.                                using=using, raw=raw)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/manager.py" in manager_method
  82.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/query.py" in _insert
  1186.         return query.get_compiler(using=using).execute_sql(return_id)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/models/sql/compiler.py" in execute_sql
  1368.                 cursor.execute(sql, params)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/backends/utils.py" in execute
  99.             return super().execute(sql, params)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/backends/utils.py" in execute
  67.         return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/backends/utils.py" in _execute_with_wrappers
  76.         return executor(sql, params, many, context)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

File "/opt/specify7/ve/lib/python3.6/site-packages/django/db/backends/mysql/base.py" in execute
  76.                 raise utils.IntegrityError(*tuple(e.args))

Exception Type: IntegrityError at /api/workbench/validate_row/1/
Exception Value: (1048, "Column 'AgentID' cannot be null")
Request information:
USER: Specifyuser object (1)

GET: No GET data

POST:
Locality Name = 'Lawrence 3'
Latitude1 = '40'
Longitude1 = '-105'
Latitude2 = '0'
Longitude2 = '0'
Lat Long Type = 'Point'
Lat Long Accuracy = '0'
Country = 'USA'
State = 'Kansas'
County = 'Douglas'

FILES: No FILES data

COOKIES:
collection = '4'
csrftoken = 'e3EWxbRpIloRLTakR982g9nfzPAhMwL43iwry6sdf2AsD2SsrKi1TJkx1Ppl8RzL'
sessionid = 'gxf3rswl34he2sz4e060woo3pfbj5lo6'

META:
CELERY_BROKER_URL = 'redis://redis/0'
CELERY_RESULT_BACKEND = 'redis://redis/1'
CONTENT_LENGTH = '158'
CONTENT_TYPE = 'application/x-www-form-urlencoded; charset=UTF-8'
CSRF_COOKIE = 'e3EWxbRpIloRLTakR982g9nfzPAhMwL43iwry6sdf2AsD2SsrKi1TJkx1Ppl8RzL'
DATABASE_HOST = 'mariadb'
DATABASE_NAME = 'specify'
DATABASE_PORT = '3306'
DJANGO_SETTINGS_MODULE = 'settings'
GATEWAY_INTERFACE = 'CGI/1.1'
HOME = '/home/specify'
HOSTNAME = '4e41659fd1b9'
HTTP_ACCEPT = '*/*'
HTTP_ACCEPT_ENCODING = 'gzip, deflate, br'
HTTP_ACCEPT_LANGUAGE = 'en-US,en;q=0.9,uk;q=0.8'
HTTP_CACHE_CONTROL = 'no-cache'
HTTP_CONNECTION = 'close'
HTTP_COOKIE = 'collection=4; csrftoken=e3EWxbRpIloRLTakR982g9nfzPAhMwL43iwry6sdf2AsD2SsrKi1TJkx1Ppl8RzL; sessionid=gxf3rswl34he2sz4e060woo3pfbj5lo6'
HTTP_HOST = 'localhost'
HTTP_ORIGIN = 'http://localhost'
HTTP_PRAGMA = 'no-cache'
HTTP_REFERER = 'http://localhost/specify/workbench/1/'
HTTP_SEC_CH_UA = '"Google Chrome";v="89", "Chromium";v="89", ";Not A Brand";v="99"'
HTTP_SEC_CH_UA_MOBILE = '?0'
HTTP_SEC_FETCH_DEST = 'empty'
HTTP_SEC_FETCH_MODE = 'cors'
HTTP_SEC_FETCH_SITE = 'same-origin'
HTTP_USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.40 Safari/537.36'
HTTP_X_CSRFTOKEN = 'e3EWxbRpIloRLTakR982g9nfzPAhMwL43iwry6sdf2AsD2SsrKi1TJkx1Ppl8RzL'
HTTP_X_FORWARDED_FOR = '172.19.0.1'
HTTP_X_REAL_IP = '172.19.0.1'
HTTP_X_REQUESTED_WITH = 'XMLHttpRequest'
LOG_LEVEL = 'WARNING'
MASTER_NAME = 'master'
MASTER_PASSWORD = 'master'
PATH = '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
PATH_INFO = '/api/workbench/validate_row/1/'
PWD = '/opt/specify7'
QUERY_STRING = ''
REMOTE_ADDR = '172.19.0.5'
REMOTE_HOST = ''
REPORT_RUNNER_HOST = 'report-runner'
REPORT_RUNNER_PORT = '8080'
REQUEST_METHOD = 'POST'
RUN_MAIN = 'true'
SCRIPT_NAME = ''
SECRET_KEY = 'change this to some unique random string'
SERVER_NAME = '4e41659fd1b9'
SERVER_PORT = '8000'
SERVER_PROTOCOL = 'HTTP/1.0'
SERVER_SOFTWARE = 'WSGIServer/0.2'
SP7_DEBUG = 'true'
TZ = 'America/Chicago'
wsgi.errors = <_io.TextIOWrapper name='<stderr>' mode='w' encoding='ANSI_X3.4-1968'>
wsgi.file_wrapper = ''
wsgi.input = <django.core.handlers.wsgi.LimitedStream object at 0x7f37247590f0>
wsgi.multiprocess = False
wsgi.multithread = True
wsgi.run_once = False
wsgi.url_scheme = 'http'
wsgi.version = '(1, 0)'

Settings:
Using settings module settings
ABSOLUTE_URL_OVERRIDES = {}
ADMINS = '()'
ADMIN_MEDIA_PREFIX = '/static/admin/'
ALLOWED_HOSTS = ['*']
ALLOW_SPECIFY6_PASSWORDS = '********************'
ALLOW_SUPPORT_LOGIN = False
ANONYMOUS_USER = None
APPEND_SLASH = True
AUTHENTICATION_BACKENDS = ['django.contrib.auth.backends.ModelBackend']
AUTH_LDAP_SERVER_URI = None
AUTH_PASSWORD_VALIDATORS = '********************'
AUTH_USER_MODEL = 'specify.Specifyuser'
CACHES = {'default': {'BACKEND': 'django.core.cache.backends.locmem.LocMemCache'}}
CACHE_MIDDLEWARE_ALIAS = 'default'
CACHE_MIDDLEWARE_KEY_PREFIX = '********************'
CACHE_MIDDLEWARE_SECONDS = 600
CELERY_BROKER_URL = 'redis://redis/0'
CELERY_RESULT_BACKEND = 'redis://redis/1'
CELERY_TASK_DEFAULT_QUEUE = 'specify'
CSRF_COOKIE_AGE = 31449600
CSRF_COOKIE_DOMAIN = None
CSRF_COOKIE_HTTPONLY = False
CSRF_COOKIE_NAME = 'csrftoken'
CSRF_COOKIE_PATH = '/'
CSRF_COOKIE_SAMESITE = 'Lax'
CSRF_COOKIE_SECURE = False
CSRF_FAILURE_VIEW = 'django.views.csrf.csrf_failure'
CSRF_HEADER_NAME = 'HTTP_X_CSRFTOKEN'
CSRF_TRUSTED_ORIGINS = []
CSRF_USE_SESSIONS = False
DATABASES = {'default': {'ENGINE': 'specifyweb.hibernateboolsbackend.backends.mysql', 'NAME': 'specify', 'USER': 'master', 'PASSWORD': '********************', 'HOST': 'mariadb', 'PORT': '3306', 'TEST': {'CHARSET': None, 'COLLATION': None, 'NAME': None, 'MIRROR': None}, 'ATOMIC_REQUESTS': False, 'AUTOCOMMIT': True, 'CONN_MAX_AGE': 0, 'OPTIONS': {}, 'TIME_ZONE': None}}
DATABASE_HOST = 'mariadb'
DATABASE_NAME = 'specify'
DATABASE_PORT = '3306'
DATABASE_ROUTERS = []
DATA_UPLOAD_MAX_MEMORY_SIZE = 2621440
DATA_UPLOAD_MAX_NUMBER_FIELDS = 1000
DATETIME_FORMAT = 'N j, Y, P'
DATETIME_INPUT_FORMATS = ['%Y-%m-%d %H:%M:%S', '%Y-%m-%d %H:%M:%S.%f', '%Y-%m-%d %H:%M', '%Y-%m-%d', '%m/%d/%Y %H:%M:%S', '%m/%d/%Y %H:%M:%S.%f', '%m/%d/%Y %H:%M', '%m/%d/%Y', '%m/%d/%y %H:%M:%S', '%m/%d/%y %H:%M:%S.%f', '%m/%d/%y %H:%M', '%m/%d/%y']
DATE_FORMAT = 'N j, Y'
DATE_INPUT_FORMATS = ['%Y-%m-%d', '%m/%d/%Y', '%m/%d/%y', '%b %d %Y', '%b %d, %Y', '%d %b %Y', '%d %b, %Y', '%B %d %Y', '%B %d, %Y', '%d %B %Y', '%d %B, %Y']
DEBUG = True
DEBUG_PROPAGATE_EXCEPTIONS = False
DECIMAL_SEPARATOR = '.'
DEFAULT_CHARSET = 'utf-8'
DEFAULT_CONTENT_TYPE = 'text/html'
DEFAULT_EXCEPTION_REPORTER_FILTER = 'django.views.debug.SafeExceptionReporterFilter'
DEFAULT_FILE_STORAGE = 'django.core.files.storage.FileSystemStorage'
DEFAULT_FROM_EMAIL = 'webmaster@localhost'
DEFAULT_INDEX_TABLESPACE = ''
DEFAULT_TABLESPACE = ''
DEPOSITORY_DIR = '/volumes/static-files/depository'
DISABLE_AUDITING = False
DISALLOWED_USER_AGENTS = []
EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
EMAIL_HOST = 'localhost'
EMAIL_HOST_PASSWORD = '********************'
EMAIL_HOST_USER = ''
EMAIL_PORT = 25
EMAIL_SSL_CERTFILE = None
EMAIL_SSL_KEYFILE = '********************'
EMAIL_SUBJECT_PREFIX = '[Django] '
EMAIL_TIMEOUT = None
EMAIL_USE_LOCALTIME = False
EMAIL_USE_SSL = False
EMAIL_USE_TLS = False
FILE_CHARSET = 'utf-8'
FILE_UPLOAD_DIRECTORY_PERMISSIONS = None
FILE_UPLOAD_HANDLERS = ['django.core.files.uploadhandler.MemoryFileUploadHandler', 'django.core.files.uploadhandler.TemporaryFileUploadHandler']
FILE_UPLOAD_MAX_MEMORY_SIZE = 2621440
FILE_UPLOAD_PERMISSIONS = None
FILE_UPLOAD_TEMP_DIR = None
FIRST_DAY_OF_WEEK = 0
FIXTURE_DIRS = []
FORCE_SCRIPT_NAME = None
FORMAT_MODULE_PATH = None
FORM_RENDERER = 'django.forms.renderers.DjangoTemplates'
IGNORABLE_404_URLS = []
INSTALLED_APPS = "('django.contrib.sessions', 'django.contrib.staticfiles', 'django.contrib.contenttypes', 'django.contrib.auth', 'specifyweb.specify', 'specifyweb.stored_queries', 'specifyweb.businessrules', 'specifyweb.express_search', 'specifyweb.context', 'specifyweb.attachment_gw', 'specifyweb.frontend', 'specifyweb.barvis', 'specifyweb.report_runner', 'specifyweb.interactions', 'specifyweb.workbench', 'specifyweb.notifications', 'specifyweb.export', 'specifyweb.raven_placeholder', 'django_jsonfield_backport')"
INTERNAL_IPS = []
JAVA_PATH = '/usr/bin/java'
LANGUAGES = [('af', 'Afrikaans'), ('ar', 'Arabic'), ('ast', 'Asturian'), ('az', 'Azerbaijani'), ('bg', 'Bulgarian'), ('be', 'Belarusian'), ('bn', 'Bengali'), ('br', 'Breton'), ('bs', 'Bosnian'), ('ca', 'Catalan'), ('cs', 'Czech'), ('cy', 'Welsh'), ('da', 'Danish'), ('de', 'German'), ('dsb', 'Lower Sorbian'), ('el', 'Greek'), ('en', 'English'), ('en-au', 'Australian English'), ('en-gb', 'British English'), ('eo', 'Esperanto'), ('es', 'Spanish'), ('es-ar', 'Argentinian Spanish'), ('es-co', 'Colombian Spanish'), ('es-mx', 'Mexican Spanish'), ('es-ni', 'Nicaraguan Spanish'), ('es-ve', 'Venezuelan Spanish'), ('et', 'Estonian'), ('eu', 'Basque'), ('fa', 'Persian'), ('fi', 'Finnish'), ('fr', 'French'), ('fy', 'Frisian'), ('ga', 'Irish'), ('gd', 'Scottish Gaelic'), ('gl', 'Galician'), ('he', 'Hebrew'), ('hi', 'Hindi'), ('hr', 'Croatian'), ('hsb', 'Upper Sorbian'), ('hu', 'Hungarian'), ('hy', 'Armenian'), ('ia', 'Interlingua'), ('id', 'Indonesian'), ('io', 'Ido'), ('is', 'Icelandic'), ('it', 'Italian'), ('ja', 'Japanese'), ('ka', 'Georgian'), ('kab', 'Kabyle'), ('kk', 'Kazakh'), ('km', 'Khmer'), ('kn', 'Kannada'), ('ko', 'Korean'), ('lb', 'Luxembourgish'), ('lt', 'Lithuanian'), ('lv', 'Latvian'), ('mk', 'Macedonian'), ('ml', 'Malayalam'), ('mn', 'Mongolian'), ('mr', 'Marathi'), ('my', 'Burmese'), ('nb', 'Norwegian Bokmål'), ('ne', 'Nepali'), ('nl', 'Dutch'), ('nn', 'Norwegian Nynorsk'), ('os', 'Ossetic'), ('pa', 'Punjabi'), ('pl', 'Polish'), ('pt', 'Portuguese'), ('pt-br', 'Brazilian Portuguese'), ('ro', 'Romanian'), ('ru', 'Russian'), ('sk', 'Slovak'), ('sl', 'Slovenian'), ('sq', 'Albanian'), ('sr', 'Serbian'), ('sr-latn', 'Serbian Latin'), ('sv', 'Swedish'), ('sw', 'Swahili'), ('ta', 'Tamil'), ('te', 'Telugu'), ('th', 'Thai'), ('tr', 'Turkish'), ('tt', 'Tatar'), ('udm', 'Udmurt'), ('uk', 'Ukrainian'), ('ur', 'Urdu'), ('vi', 'Vietnamese'), ('zh-hans', 'Simplified Chinese'), ('zh-hant', 'Traditional Chinese')]
LANGUAGES_BIDI = ['he', 'ar', 'fa', 'ur']
LANGUAGE_CODE = 'en-us'
LANGUAGE_COOKIE_AGE = None
LANGUAGE_COOKIE_DOMAIN = None
LANGUAGE_COOKIE_NAME = 'django_language'
LANGUAGE_COOKIE_PATH = '/'
LOCALE_PATHS = []
LOGGING = {'version': 1, 'disable_existing_loggers': False, 'formatters': {'standard': {'format': '[%(asctime)s] [%(levelname)s] [%(name)s:%(lineno)s] %(message)s', 'datefmt': '%d/%b/%Y %H:%M:%S'}}, 'handlers': {'console': {'level': 'WARNING', 'class': 'logging.StreamHandler', 'formatter': 'standard'}}, 'loggers': {'specifyweb': {'handlers': ['console'], 'level': 'DEBUG', 'propagate': True}}}
LOGGING_CONFIG = 'logging.config.dictConfig'
LOGIN_REDIRECT_URL = '/'
LOGIN_URL = '/accounts/login/'
LOGOUT_REDIRECT_URL = None
MANAGERS = '()'
MASTER_NAME = 'master'
MASTER_PASSWORD = '********************'
MEDIA_ROOT = ''
MEDIA_URL = ''
MESSAGE_STORAGE = 'django.contrib.messages.storage.fallback.FallbackStorage'
MIDDLEWARE = ['django.middleware.gzip.GZipMiddleware', 'django.middleware.common.CommonMiddleware', 'django.contrib.sessions.middleware.SessionMiddleware', 'django.middleware.csrf.CsrfViewMiddleware', 'django.contrib.auth.middleware.AuthenticationMiddleware', 'specifyweb.context.middleware.ContextMiddleware']
MIGRATION_MODULES = {}
MONTH_DAY_FORMAT = 'F j'
NOTIFICATION_TTL_DAYS = 7
NUMBER_GROUPING = 0
PASSWORD_HASHERS = '********************'
PASSWORD_RESET_TIMEOUT_DAYS = '********************'
PREPEND_WWW = False
RAVEN_CONFIG = None
REPORT_RUNNER_HOST = 'report-runner'
REPORT_RUNNER_PORT = '8080'
ROOT_URLCONF = 'specifyweb.urls'
RO_MODE = False
SA_DATABASE_URL = 'mysql://master:master@mariadb:3306/specify?charset=utf8'
SA_POOL_RECYCLE = 3600
SCHEMA_LANGUAGE = 'en'
SECRET_KEY = '********************'
SECURE_BROWSER_XSS_FILTER = False
SECURE_CONTENT_TYPE_NOSNIFF = False
SECURE_HSTS_INCLUDE_SUBDOMAINS = False
SECURE_HSTS_PRELOAD = False
SECURE_HSTS_SECONDS = 0
SECURE_PROXY_SSL_HEADER = None
SECURE_REDIRECT_EXEMPT = []
SECURE_SSL_HOST = None
SECURE_SSL_REDIRECT = False
SERVER_EMAIL = 'root@localhost'
SESSION_CACHE_ALIAS = 'default'
SESSION_COOKIE_AGE = 1209600
SESSION_COOKIE_DOMAIN = None
SESSION_COOKIE_HTTPONLY = True
SESSION_COOKIE_NAME = 'sessionid'
SESSION_COOKIE_PATH = '/'
SESSION_COOKIE_SAMESITE = 'Lax'
SESSION_COOKIE_SECURE = False
SESSION_ENGINE = 'django.contrib.sessions.backends.file'
SESSION_EXPIRE_AT_BROWSER_CLOSE = True
SESSION_FILE_PATH = None
SESSION_SAVE_EVERY_REQUEST = False
SESSION_SERIALIZER = 'django.contrib.sessions.serializers.JSONSerializer'
SETTINGS_MODULE = 'settings'
SHORT_DATETIME_FORMAT = 'm/d/Y P'
SHORT_DATE_FORMAT = 'm/d/Y'
SIGNING_BACKEND = 'django.core.signing.TimestampSigner'
SILENCED_SYSTEM_CHECKS = []
SITE_ID = 1
SPECIFY_CONFIG_DIR = '/opt/Specify/config'
SPECIFY_THICK_CLIENT = '/opt/Specify'
STATICFILES_DIRS = "(('config', '/opt/Specify/config'),)"
STATICFILES_FINDERS = "('django.contrib.staticfiles.finders.FileSystemFinder', 'django.contrib.staticfiles.finders.AppDirectoriesFinder')"
STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.StaticFilesStorage'
STATIC_ROOT = ''
STATIC_URL = '/static/'
STATS_URL = 'https://stats.specifycloud.org/capture'
SUPPORT_LOGIN_TTL = 300
TEMPLATES = [{'BACKEND': 'django.template.backends.django.DjangoTemplates', 'DIRS': [], 'APP_DIRS': True, 'OPTIONS': {'context_processors': ['django.contrib.auth.context_processors.auth', 'django.template.context_processors.debug', 'django.template.context_processors.i18n', 'django.template.context_processors.media', 'django.template.context_processors.static', 'django.template.context_processors.tz', 'django.contrib.messages.context_processors.messages']}}]
TEST_NON_SERIALIZED_APPS = []
TEST_RUNNER = 'django.test.runner.DiscoverRunner'
THICK_CLIENT_LOCATION = '/opt/Specify'
THOUSAND_SEPARATOR = ','
TIME_FORMAT = 'P'
TIME_INPUT_FORMATS = ['%H:%M:%S', '%H:%M:%S.%f', '%H:%M']
TIME_ZONE = 'America/Chicago'
USE_I18N = True
USE_L10N = True
USE_THOUSAND_SEPARATOR = False
USE_TZ = False
USE_X_FORWARDED_HOST = False
USE_X_FORWARDED_PORT = False
VERSION = '(debug)'
WB_UPLOAD_LOG_DIR = '/home/specify/wb_upload_logs'
WEB_ATTACHMENT_COLLECTION = None
WEB_ATTACHMENT_KEY = '********************'
WEB_ATTACHMENT_REQUIRES_KEY_FOR_GET = '********************'
WEB_ATTACHMENT_URL = None
WSGI_APPLICATION = None
X_FRAME_OPTIONS = 'SAMEORIGIN'
YEAR_MONTH_FORMAT = 'F Y'


You're seeing this error because you have DEBUG = True in your
Django settings file. Change that to False, and Django will
display a standard page generated by the handler for this status code.

```

**Comment by maxxxxxdlp** [comment]
Still happening
@benanhalt  Should I go ahead and prevent uploading if not all required fields were mapped? Or would you fix it on the backend?

**Comment by benanhalt** [comment]
I need to add an exception handler for the IntegrityErrors anyways.

**Comment by maxxxxxdlp** [comment]
Now I get the message `(1048, "Column 'AgentID' cannot be null")` when I hover over an invalid cell
The message is not very intuitive (it doesn't say that the Agent is missing in plain English), but still, better than throwing an exception

**Comment by maxxxxxdlp** [comment]
@benanhalt would it be hard to improve the error message? Maybe specify the table name in it and check whether it is a relationship, e.x `Relationship with the "Agent" table is mandatory and needs to be specified` or even better: `Please map the "Agent" table before uploading`

**Comment by maxxxxxdlp** [comment]
Found another case where missing required field causes cryptic messages

![Screen Shot 2021-06-26 at 4 36 43 PM](https://user-images.githubusercontent.com/40512816/123514647-b2ec3980-d69c-11eb-963f-f1ce35da9be7.png)

Data set:
```json
{
  "id": 13,
  "name": "New Data Set Sat Jun 26 2021",
  "columns": [
    "Agent Last Name",
    "Agent Last Name #2",
    "Date Accessioned"
  ],
  "visualorder": [
    0,
    1,
    2
  ],
  "rows": [
    [
      "1",
      "1",
      "3",
      "{\"disambiguation\":{}}"
    ]
  ],
  "uploadplan": {
    "baseTableName": "accession",
    "uploadable": {
      "uploadTable": {
        "wbcols": {
          "dateaccessioned": "Date Accessioned"
        },
        "static": {},
        "toOne": {},
        "toMany": {
          "accessionagents": [
            {
              "wbcols": {},
              "static": {},
              "toOne": {
                "agent": {
                  "uploadTable": {
                    "wbcols": {
                      "lastname": "Agent Last Name"
                    },
                    "static": {},
                    "toOne": {},
                    "toMany": {}
                  }
                }
              }
            },
            {
              "wbcols": {},
              "static": {},
              "toOne": {
                "agent": {
                  "uploadTable": {
                    "wbcols": {
                      "lastname": "Agent Last Name #2"
                    },
                    "static": {},
                    "toOne": {},
                    "toMany": {}
                  }
                }
              }
            }
          ]
        }
      }
    }
  },
  "uploaderstatus": null,
  "uploadresult": null,
  "rowresults": [
    {
      "UploadResult": {
        "record_result": {
          "Uploaded": {
            "id": 45,
            "info": {
              "tableName": "Accession",
              "columns": [
                "Date Accessioned"
              ],
              "treeInfo": null
            },
            "picklistAdditions": []
          }
        },
        "toOne": {},
        "toMany": {
          "accessionagents": [
            {
              "UploadResult": {
                "record_result": {
                  "Uploaded": {
                    "id": 5,
                    "info": {
                      "tableName": "Accessionagent",
                      "columns": [],
                      "treeInfo": null
                    },
                    "picklistAdditions": []
                  }
                },
                "toOne": {
                  "agent": {
                    "UploadResult": {
                      "record_result": {
                        "Matched": {
                          "id": 160,
                          "info": {
                            "tableName": "Agent",
                            "columns": [
                              "Agent Last Name"
                            ],
                            "treeInfo": null
                          }
                        }
                      },
                      "toOne": {},
                      "toMany": {}
                    }
                  }
                },
                "toMany": {}
              }
            },
            {
              "UploadResult": {
                "record_result": {
                  "FailedBusinessRule": {
                    "message": "(1062, \"Duplicate entry '160-45' for key 'Role'\")",
                    "info": {
                      "tableName": "Accessionagent",
                      "columns": [],
                      "treeInfo": null
                    }
                  }
                },
                "toOne": {
                  "agent": {
                    "UploadResult": {
                      "record_result": {
                        "Matched": {
                          "id": 160,
                          "info": {
                            "tableName": "Agent",
                            "columns": [
                              "Agent Last Name #2"
                            ],
                            "treeInfo": null
                          }
                        }
                      },
                      "toOne": {},
                      "toMany": {}
                    }
                  }
                },
                "toMany": {}
              }
            }
          ]
        }
      }
    }
  ],
  "remarks": null,
  "importedfilename": "",
  "timestampcreated": "2021-06-26T10:40:39.416",
  "timestampmodified": "2021-06-26T10:52:56.763",
  "createdbyagent": "/api/specify/agent/1/",
  "modifiedbyagent": "/api/specify/agent/1/"
}
```

**Comment by maxxxxxdlp** [comment]
The original issue has been fixed. Missing required relationships now have a nice error message.

Still, missing numbering fields (cat #, accession #, etc) cause ugly error messages.

See https://github.com/specify/specify7/issues/777#issuecomment-869003850

**Comment by benanhalt** [comment]
This particular error is because (I think) the field is not required explicitly required by Specify or the database but has a uniqueness constraint in the database. If there are multiple rows with no value, they would get assigned null, which is not unique.

**Comment by benanhalt** [comment]
I'm going to mark this as fixed and you can create a separate issue for the new problem.

**Comment by maxxxxxdlp** [comment]
Created #872
